api/odilo.py

```diff
@@ -0,0 +1,532 @@
+
+        if code == 'HOLD_NOT_FOUND':
+            return True
+
+        raise CannotReleaseHold(description)
+
+
+class OdiloCirculationMonitor(CollectionMonitor):
+    """Maintain LicensePools for recently changed Odilo titles
+    """
+    SERVICE_NAME = "Odilo Circulation Monitor"
+    INTERVAL_SECONDS = 500
+    PROTOCOL = ExternalIntegration.ODILO
+
+    def __init__(self, _db, collection, api_class=OdiloAPI):
+        """Constructor."""
+        super(OdiloCirculationMonitor, self).__init__(_db, collection)
+        self.api = api_class(_db, collection)
+
+    def run_once(self, start, cutoff):
+        self.log.info("Starting recently_changed_ids, start: " + str(start) + ", cutoff: " + str(cutoff))
+
+        start_time = datetime.datetime.now()
+        self.recently_changed_ids(start)
+        finish_time = datetime.datetime.now()
+
+        time_elapsed = finish_time - start_time
+        self.log.info("recently_changed_ids finished in: " + str(time_elapsed))
+
+    def recently_changed_ids(self, start):
+        modification_date = None
+        if start:
+            if isinstance(start, datetime.date):
+                modification_date = start.strftime('%Y-%m-%d')  # Format YYYY-MM-DD
+            elif isinstance(start, basestring):
+                modification_date = start
+
+        self.all_ids(modification_date)
+
+    def all_ids(self, modication_date=None):
+        """Get IDs for every book in the system, from modification date if any
+        """
+
+        retrieved = 0
+        parsed = 0
+        new = 0
+        offset = 0
+        limit = self.api.PAGE_SIZE_LIMIT
+
+        url = "%s?limit=%i&offset=%i" % (self.api.ALL_PRODUCTS_ENDPOINT, limit, offset)
+        if modication_date:
+            url = "%s&modificationDate=%s" % (url, modication_date)
+
+        # Retrieve first group of records
+        status_code, headers, content = self.api.get(url)
+        content = json.loads(content)
+
+        # Retrieve Odilo record in groups
+        while status_code == 200 and len(content) > 0:
+            offset += limit
+            retrieved += len(content)
+            self.log.info('Retrieved %i records' % retrieved)
+
+            # Process a bunch of records retrieved
+            for record in content:
+                record_id = record['id']
+                self.log.info('Processing record %i/%i: %s' % (parsed, retrieved, record_id))
+                identifier, is_new = self.api.odilo_bibliographic_coverage_provider.process_item(
+                    record_id, record
+                )
+
+                if is_new:
+                    new += 1
+
+                parsed += 1
+
+            # Persist each bunch of retrieved records
+            self._db.commit()
+
+            # Retrieve next group of records
+            url = "%s?limit=%i&offset=%i" % (self.api.ALL_PRODUCTS_ENDPOINT, limit, offset)
+            if modication_date:
+                url = "%s&modificationDate=%s" % (url, modication_date)
```

Comment:
You have this code twice and its complexity makes it a good candidate for refactoring.

Reply:
Ok, I refactorized it.